### PR TITLE
cleanup for lookup EOS

### PR DESCRIPTION
**Pull request by jdannberg**
This makes a few changes to the lookup equation of state (see discussion in  #3828 and #4148)

Primarily, it moves the fill mass and volume fraction functionality back into the material model, using an existing utility function rather than the Steinberger model using its own. This includes checking if the model has a background field: The material model should interpret what the compositional fields mean, not the equation of state, that just fills outputs for individual compositions. This change also makes it easier to extend this in the future to be able to handle compositional fields that do not represent mass fractions. 

This also changes the new fill_averaged_equation_of_state_outputs function from #4148 to only average properties in one location, not in all points in the cell. That is both more convenient for this material model, but also important if the functionality is to be used in other material models: They all just have one EquationOfStateOutputs object, not a vector of them (see for example the multicomponent material model). 

This does not yet solve the problem where to move that function; I think the equation of state interface is not the ideal place for it, but @bobmyhill is right, we can also not easily move it to the utilities. 

I have tried in this pull request to recreate the exact way this averaging worked in the Steinberger model before, but a few of the tests fail for me locally (with small deviations from the previous results), so I also want to see what the tester says to make sure I haven't changed any functionality. 

* [x] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

### CODE REVIEW

**Review by bobmyhill — CHANGES_REQUESTED** [review]
Looks great, thanks!

I don't have any suggestions for the changes that you've made, but there are two additions which would be good to include in this PR:
- Declare parameters in thermodynamic_table_lookup.cc still has "Index of first mass fraction compositional field", which should be removed (oops, sorry!).
- I don't think there's a test for when the number of lookups = 1 but the number of compositional fields > 1. Could you add one, so that we can't break any user code?

**Review by bobmyhill — APPROVED** [review]

**Comment by jdannberg** [comment]
I see, very clever :grin: 
Should be fixed now.

File: include/aspect/material_model/utilities.h [hunk 1]
```diff
@@ -242,7 +242,8 @@ namespace aspect
        */
       std::vector<double>
       compute_composition_fractions(const std::vector<double> &compositional_fields,
-                                    const ComponentMask &field_mask = ComponentMask());
+                                    const ComponentMask &field_mask = ComponentMask(),
+                                    const bool has_background_field = true);
```
Comment (gassmoeller) [inline include/aspect/material_model/utilities.h:246]: Document this parameter and that it changes the size of the returned vector.

**Review by gassmoeller — COMMENTED** [review]
I am late for this review, but maybe you are willing to do some fixes in a follow-up PR? I think in general this is correct, but I have a few suggestions to make the code easier to read, and a few questions about the functionality.

File: source/material_model/equation_of_state/thermodynamic_table_lookup.cc [hunk 2]
```diff
@@ -440,15 +389,6 @@ namespace aspect
                            "List with as many components as active "
                            "compositional fields (material data is assumed to "
                            "be in order with the ordering of the fields).");
-        prm.declare_entry ("Index of first mass fraction compositional field", "0",
-                           Patterns::Integer (0),
-                           "The index of the first compositional field which "
-                           "corresponds to the mass fraction of a material. "
-                           "The indexing starts at 0.");
```
[the remark this answers is not in the document]
Comment (jdannberg) [reply source/material_model/equation_of_state/thermodynamic_table_lookup.cc:447]: Yes.

**Pull request merged**